### PR TITLE
windows: add support for device query requests memstat|memstat_raw|mem_group_topology

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Added device queries memstat|memstat_raw|mem_group_topology which are
required for the memory information output in "xbutil examine -r memory"

Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
Windows device driver support is not ready yet, so added dummy changes in the device driver get a output for "xbutil examine **memory** command"

```
xbutil examine -r memory

-----------------------------------------------
1/1 [0000:00:00.0] : xilinx_u250_xdma_201830_2
-----------------------------------------------
Memory Information

  Memory Topology
         Tag         Type        Temp(C)  Size      Base Address
    [ 0] bank0       MEM_DDR4    212      16 GB     0x4000000000
    [ 1] bank1       MEM_DDR4    N/A      16 GB     0x5000000000
    [ 2] bank2       MEM_DDR4    N/A      16 GB     0x6000000000
    [ 3] bank3       MEM_DDR4    N/A      16 GB     0x7000000000

  Memory Status
         Tag         Type        Size    Mem Usage       BO count
    [ 0] bank0       MEM_DDR4    16 GB   1001 Byte       1
    [ 1] bank1       MEM_DDR4    16 GB   2001 Byte       2
    [ 2] bank2       MEM_DDR4    16 GB   3001 Byte       3
    [ 3] bank3       MEM_DDR4    16 GB   4001 Byte       4

```

File: src/runtime_src/core/pcie/windows/shim.cpp [hunk 1]
```diff
@@ -903,7 +903,7 @@ struct shim
     if (!status || bytes != sizeof(struct mem_topology))
       throw std::runtime_error("DeviceIoControl IOCTL_XOCL_STAT (get_mem_topology) failed");
 
-    DWORD mem_topology_size = sizeof(struct mem_topology) + (mem_info.m_count - 1) * sizeof(struct mem_data);
+    DWORD mem_topology_size = sizeof(struct mem_topology) + (mem_info.m_count) * sizeof(struct mem_data);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/shim.cpp:906]: I think the point is that sizeof(mem_topology) already contains the size of one mem_data.  This is indeed an implicit assumption that if true should be mentioned in a comment prior to the original line here.

I believe same is true for the other xclbin data types you size up below.

**Review by stsoe — COMMENTED** [review]
Looks good.  One comment related to sizing up the xclbin structs.

File: src/runtime_src/core/pcie/windows/shim.cpp (hunk 1 above)
Comment (rajkumar-xilinx) [reply src/runtime_src/core/pcie/windows/shim.cpp]: True, fixed this, thank you.

**Comment by rajkumar-xilinx** [comment]
`>xbutil examine -r all
System Configuration
  OS Name              : Windows NT
  Release              : 20215.rs_prerelease.200911-1500
  Version              : 6.3
  Machine              : x86_64
  CPU Cores            : 6
  Memory               : 16382 MB
  Distribution         : Windows Server 2019 Datacenter
  Model                : Virtual Machine

XRT
  Version              : 2.12.0
  Branch               : 2aug
  Hash                 : 4dd460d0d20774aa4335b295f5770d7a32c2afa8
  Hash Date            : 2021-08-03 09:19:42

Devices present
  [0000:00:00.0] : xilinx_u250_xdma_201830_2


-----------------------------------------------
1/1 [0000:00:00.0] : xilinx_u250_xdma_201830_2
-----------------------------------------------
AIE
  <AIE information unavailable>

Memory Information

  Memory Topology
         Tag         Type        Temp(C)  Size      Base Address
    [ 0] bank0       MEM_DDR4    212      16 GB     0x4000000000
    [ 1] bank1       MEM_DDR4    N/A      16 GB     0x5000000000
    [ 2] bank2       MEM_DDR4    N/A      16 GB     0x6000000000
    [ 3] bank3       MEM_DDR4    N/A      16 GB     0x7000000000

  Memory Status
         Tag         Type        Size    Mem Usage       BO count
    [ 0] bank0       MEM_DDR4    16 GB   1001 Byte       1
    [ 1] bank1       MEM_DDR4    16 GB   2001 Byte       2
    [ 2] bank2       MEM_DDR4    16 GB   3001 Byte       3
    [ 3] bank3       MEM_DDR4    16 GB   4001 Byte       4

Xclbin UUID
  C5B9A584-9B70-4902-AE32-ADDF5C1C6E0C

Compute Units
  PL Compute Units
    Index   Name                          Base_Address    Usage   Status
    0       bandwidth1:kernel_1           0x1800000       2       (--)
    1       bandwidth2:kernel_2           0x1810000       2       (--)

  PS Compute Units
    Index   Name                          Base_Address    Usage   Status

Debug IP Status
  INFO: Failed to find any Debug IP Layout section in the bitstream loaded on device. Ensure that a valid bitstream with debug IPs (AIM, LAPC) is successfully downloaded.

 INFO: Debug IP Data is not populated.
 INFO: Debug IP Data is not populated.
Pcie Info
  Vendor                 : 0x10ee
  Device                 : 0x5005
  Sub Device             : 0x000e
  Sub Vendor             : 0x10ee
  PCIe                   : Gen3x16
  DMA Thread Count       : 0
  CPU Affinity           : 0
  Shared Host Memory     : 0 Bytes
  Max Shared Host Memory : 0 Bytes

Platform
  XSA Name               : xilinx_u250_xdma_201830_2
  FPGA Name              : xcu250-figd2104-2L-e
  JTAG ID Code           : 0x4b57093
  DDR Size               : 68719476736 Bytes
  DDR Count              : 4
  Mig Calibrated         : true
  P2P Status             : not supported

Mac Addresses            : 00:0A:35:05:FE:28
                         : 00:0A:35:05:FE:29

Electrical
  Max Power              : 75 Watts
  Power                  : 44.398284 Watts
  Power Warning          : false

  Power Rails            : Voltage   Current
  12 Volts Auxillary     :  0.471 V,  0.020 A
  12 Volts PCI Express   : 12.168 V,  3.648 A
  3.3 Volts PCI Express  :  3.353 V
  3.3 Volts Auxillary    :  3.304 V
  Internal FPGA Vcc      :  0.851 V, 13.422 A
  DDR Vpp Bottom         :  2.500 V
  DDR Vpp Top            :  2.500 V
  5.5 Volts System       :  5.494 V
  Vcc 1.2 Volts Top      :  1.211 V
  Vcc 1.2 Volts Bottom   :  1.201 V
  1.8 Volts Top          :  1.828 V
  0.9 Volts Vcc          :  0.912 V
  12 Volts SW            : 12.139 V
  Mgt Vtt                :  1.204 V

Mailbox
  Total bytes received   : 3904 Bytes
  Unknown                : 0
  Test msg ready         : 0
  Test msg fetch         : 0
  Lock bitstream         : 0
  Unlock bitstream       : 0
  Hot reset              : 0
  Firewall trip          : 0
  Download xclbin kaddr  : 0
  Download xclbin        : 0
  Reclock                : 0
  Peer data read         : 0
  User probe             : 0
  Mgmt state             : 0
  Change shell           : 0
  Reprogram shell        : 0
  P2P bar addr           : 0

Mechanical
  Fans
    FPGA Fan 1
      Critical Trigger Temp : 46 C
      Speed                 : 1260 RPM

Firewall
  Level 0: 0x0 (GOOD)

Thermals
  PCB Top Front          : 46 C
  PCB Top Rear           : 36 C
  PCB Bottom Front       : 44 C
  FPGA                   : 50 C

Qspi-status
  ERROR: query request (47) not supported for userpf on windows`

**Comment by stsoe** [comment]
@rajkumar-xilinx It would be good if you fixed the other size computations also and if you added a comment about the -1.

**Comment by rajkumar-xilinx** [comment]
> @rajkumar-xilinx It would be good if you fixed the other size computations also and if you added a comment about the -1.

Fixed the size computations in other places also. Thank you

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good. Please add a comment about "-1" suggested by Soren sothat it will be clear.

**Comment by dayeh-xilinx** [comment]
retest this please. u280 test got a "cold reboot timeout" retesting.

**Comment by rajkumar-xilinx** [comment]
Hi @stsoe , I have added comment in the code as suggested by you. Please review.

**Review by stsoe — APPROVED** [review]

**Pull request merged**